File: BOJ/Gold III/BOJ1516.py
#BOJ1516 게임 개발 20210406
import sys
from collections import deque
from copy import deepcopy
input = sys.stdin.readline

def topologicalSort(n, indegree, adj):
    visited = [False] * n
    q = deque()
    for i in range(n):
        if indegree[i] == 0:
            q.append(i)
            visited[i] = True
    ret = []
    while q:
        v = q.popleft()
        ret.append(v)
        for nv in adj[v]:
            indegree[nv] -= 1
            if not visited[nv] and indegree[nv] == 0:
                q.append(nv)
                visited[nv] = True
    return ret
def main():
    n = int(input())
    adj = [[] for _ in range(n)]
    indegree = [0] * n
    time = [0] * n
    for i in range(n):
        li = list(map(int,input().rstrip().split()))
        time[i] = li[0]
        for v in li[1:]:
            if v != -1:
                adj[v-1].append(i)
                indegree[i] += 1
    topology = topologicalSort(n, indegree, adj)
    res = deepcopy(time)
    for v in topology:
        for nv in adj[v]:
            res[nv] = max(res[v] + time[nv], res[nv])
    for v in res:
        print(v)
if __name__ == '__main__':
    main()

File: BOJ/Gold III/BOJ1516.java
//BOJ1516 게임 개발 20210712
import java.io.*;
import java.util.*;
import java.util.stream.Collectors;

import static java.lang.Math.*;

public class Main {
    private static final BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    private static int n,m;
    private static int[] inDegree;
    private static int[] time;
    private static int[] ans;
    private static List<List<Integer>> adj = new ArrayList<>();

    public static void main(String[] args) throws IOException {
        n = Integer.parseInt(br.readLine());

        inDegree = new int[n+1];
        time = new int[n+1];
        ans = new int[n+1];
        for(int i = 0; i <= n; i++){
            adj.add(new ArrayList<>());
        }

        for(int i = 1; i <= n; i++){
            StringTokenizer stk = new StringTokenizer(br.readLine());
            time[i] = Integer.parseInt(stk.nextToken());
            ans[i] = time[i];
            while (true){
                int from = Integer.parseInt(stk.nextToken());
                if(from == -1) break;
                inDegree[i]++;
                adj.get(from).add(i);
            }
        }
        Queue<Integer> q = new LinkedList<>();
        List<Integer> topology = new ArrayList<>();
        for(int i = 1; i <= n; i++){
            if(inDegree[i] == 0){
                q.add(i);
                topology.add(i);
            }
        }
        while(!q.isEmpty()){
            int v = q.poll();
            for(int nv: adj.get(v)){
                inDegree[nv]--;
                if(inDegree[nv] == 0){
                    q.add(nv);
                    topology.add(nv);
                }
            }
        }
        for(int v: topology){
            for(int nv: adj.get(v)){
                ans[nv] = max(ans[v] + time[nv], ans[nv]);
            }

        }
        for(int i = 1; i <= n; i++){
            bw.write(ans[i] + "\n");
        }
        bw.flush();
        bw.close();
    }
}